tt_metal/impl/device/device_pool.cpp

```diff
@@ -33,55 +33,86 @@ std::unordered_map<int, std::vector<uint32_t>> get_cpu_cores_per_numa_node(std::
         }
     } else {
-        // System does not use NUMA. Use-round robin binding strategy.
-        core_assigned_to_device = mmio_controlled_device_id % sysconf(_SC_NPROCESSORS_ONLN);
+        // NUMA node reported by UMD does not exist on host. Use round-robin binding policy for this worker thread.
+        log_warning(
+            tt::LogMetal,
+            "NUMA node {} for device {} does not exist on host or NUMA is not available.",
+            numa_node_for_device,
+            mmio_controlled_device_id);
+        core_assigned_to_device_worker_thread = mmio_controlled_device_id % num_online_processors;
+        if (use_separate_procs) {
+            core_assigned_to_device_completion_queue_reader = (mmio_controlled_device_id + num_devices) % num_online_processors;
+        } else {
+            core_assigned_to_device_completion_queue_reader = core_assigned_to_device_worker_thread;
+        }
+    }
+
+    free_cores.erase(core_assigned_to_device_worker_thread);
+    if (core_assigned_to_device_completion_queue_reader != core_assigned_to_device_worker_thread) {
+        free_cores.erase(core_assigned_to_device_completion_queue_reader);
     }
-    free_cores.erase(core_assigned_to_device);
-    return core_assigned_to_device;
+    return std::make_pair(core_assigned_to_device_worker_thread, core_assigned_to_device_completion_queue_reader);
 }
 
 std::unordered_map<uint32_t, uint32_t> get_device_id_to_core_map(
     const std::vector<chip_id_t>& device_ids,
     std::unordered_set<uint32_t>& free_cores,
-    bool use_numa_node_based_thread_binding) {
-    std::unordered_map<uint32_t, uint32_t> device_to_core_map = {};
+    bool use_numa_node_based_thread_binding,
+    const uint8_t num_hw_cqs,
+    std::unordered_map<uint32_t, uint32_t>& completion_queue_reader_to_cpu_core_map) {
+    uint32_t num_online_processors = sysconf(_SC_NPROCESSORS_ONLN);
+    constexpr uint32_t max_num_procs_per_device = 2;
+    // When using multiple command queues, assign separate CPU cores to worker and completion queue reader threads,
+    // if enough processors exist on host. Atleast one core is given to the main thread.
+    bool separate_procs_for_worker_and_reader = (num_hw_cqs > 1) && (max_num_procs_per_device * device_ids.size() <= num_online_processors - 1);
+    std::unordered_map<uint32_t, uint32_t> worker_thread_to_cpu_core_map = {};
     if (use_numa_node_based_thread_binding) {
         auto cpu_cores_per_numa_node = device_cpu_allocator::get_cpu_cores_per_numa_node(free_cores);
         for (const auto& device_id : device_ids) {
-            device_to_core_map.insert(
-                {device_id,
-                 device_cpu_allocator::get_cpu_core_for_device_worker_thread(
-                     device_id, cpu_cores_per_numa_node, free_cores)});
+            auto [worker_thread_core, completion_queue_reader_core] = device_cpu_allocator::get_cpu_core_for_device_worker_thread(
```

Comment:
nit: I think `if (use_numa_node_based_thread_binding)` and the round robin cpu assignment case could e consolidated a bit if the condition is checked inside one loop over device ids.

Reply:
I tried doing this, it gets a little messy. `device_cpu_allocator::get_cpu_cores_per_numa_node(free_cores)` is currently called in the top branch, if I consolidate the 2 branches it'll always get called, and print a warning if numa is not enabled.